cmd/swapd/contract.go

```diff
@@ -98,6 +86,82 @@ func getSwapFactory(client *ethclient.Client, addr ethcommon.Address) (*contract
 	return contracts.NewSwapFactory(addr, client)
 }
 
-func deploySwapFactory(client *ethclient.Client, txOpts *bind.TransactOpts) (ethcommon.Address, *ethtypes.Transaction, *contracts.SwapFactory, error) { //nolint:lll
-	return contracts.DeploySwapFactory(txOpts, client)
+func deploySwapFactory(
+	ctx context.Context,
+	ec *ethclient.Client,
+	privkey *ecdsa.PrivateKey,
+	forwarderAddress ethcommon.Address,
+	dataDir string,
+) (ethcommon.Address, *contracts.SwapFactory, error) {
+	if privkey == nil {
+		return ethcommon.Address{}, nil, errNoEthereumPrivateKey
+	}
+
+	chainID, err := ec.ChainID(ctx)
+	if err != nil {
+		return ethcommon.Address{}, nil, err
+	}
+	txOpts, err := bind.NewKeyedTransactorWithChainID(privkey, chainID)
+	if err != nil {
+		return ethcommon.Address{}, nil, fmt.Errorf("failed to make transactor: %w", err)
+	}
+
+	if (forwarderAddress == ethcommon.Address{}) {
+		// deploy forwarder contract as well
+		address, err := deployForwarder(ctx, ec, txOpts) //nolint:govet
+		if err != nil {
+			return ethcommon.Address{}, nil, err
+		}
+
+		forwarderAddress = address
+	}
+
+	// deploy contracts.sol
+	address, tx, sf, err := contracts.DeploySwapFactory(txOpts, ec, forwarderAddress)
+	if err != nil {
+		return ethcommon.Address{}, nil, fmt.Errorf("failed to deploy swap factory: %w", err)
+	}
+
+	_, err = block.WaitForReceipt(ctx, ec, tx.Hash())
+	if err != nil {
+		return ethcommon.Address{}, nil, err
+	}
+
+	log.Infof("deployed SwapFactory.sol: address=%s tx hash=%s", address, tx.Hash())
+
+	// store the contract address on disk
+	fp := path.Join(dataDir, "contract-address.json")
+	if err = pcommon.WriteContractAddressToFile(fp, address.String()); err != nil {
+		return ethcommon.Address{}, nil, fmt.Errorf("failed to write contract address to file: %w", err)
+	}
+
+	return address, sf, nil
+}
+
+func deployForwarder(
+	ctx context.Context,
+	ec *ethclient.Client,
+	txOpts *bind.TransactOpts,
+) (ethcommon.Address, error) {
+	address, tx, contract, err := gsnforwarder.DeployForwarder(txOpts, ec)
+	if err != nil {
+		return ethcommon.Address{}, fmt.Errorf("failed to deploy Forwarder.sol: %w", err)
+	}
+
+	_, err = block.WaitForReceipt(ctx, ec, tx.Hash())
+	if err != nil {
+		return ethcommon.Address{}, err
+	}
+
+	tx, err = contract.RegisterDomainSeparator(txOpts, gsnforwarder.DefaultName, gsnforwarder.DefaultVersion)
+	if err != nil {
+		return ethcommon.Address{}, fmt.Errorf("failed to register domain separator: %w", err)
+	}
+
+	_, err = block.WaitForReceipt(ctx, ec, tx.Hash())
+	if err != nil {
+		return ethcommon.Address{}, err
+	}
+
+	return address, nil
```

Comment:
Would it make sense to move some of the above 2 functions to the ethereum package?  Maybe not the part that is writing the contract address to disk, but the rest of it seems like it might be a good fit for that package.